### PR TITLE
fix(runtime): share deadline helper for retries and polling

### PR DESCRIPTION
**Pull request by teng-lin**
## Summary
- Add `RuntimeDeadline` as the shared internal helper for monotonic deadlines, remaining budget, sleep clamps, and timeout diagnostics.
- Wire retry sleeps to the existing client timeout budget and fail fast when a retry sleep cannot fit before the aggregate deadline.
- Move source/artifact polling sleep clamps onto the shared helper while preserving artifact poll dedupe behavior.

## Task
Phase 2 Wave 1: `p2-runtime-deadline-helper` from `.sisyphus/phases/codebase-gaps-remediation-plan/phase-2.md`.

## Wave 2 Documentation Mapping Note
Document `src/notebooklm/_deadline.py` in the architecture freshness pass as the narrow runtime deadline/sleep-clamp helper shared by retry middleware, artifact polling, and source polling.

## Test plan
- [x] `uv run pytest tests/unit/test_retry_middleware.py tests/unit/test_retry_after.py tests/unit/test_rate_limit_retry.py tests/unit/test_artifact_rate_limit_retry.py tests/unit/test_artifacts_polling_retries.py tests/unit/test_polling_service.py tests/unit/test_source_polling_service.py -q`
- [x] `uv run pytest tests/integration/concurrency/test_artifact_poll_dedupe.py -q`
- [x] `uv run ruff check src/notebooklm/_middleware_retry.py src/notebooklm/_artifact_polling.py src/notebooklm/_source_polling.py`
- [x] `uv run mypy src/notebooklm`
- [x] `uv run pytest tests/unit/test_middleware_chain_builder.py -q`

## Deferred polish findings
None.

## Summary by CodeRabbit

* **Bug Fixes**
  * Timeouts now use a unified deadline so sleep/backoff durations are clamped to remaining time, avoiding spurious waits and ensuring retries stop when time is exhausted.
  * Retry behavior preserves original errors when aggregate retry deadlines are hit and provides clearer timeout diagnostics.

* **Tests**
  * Added deterministic tests covering timeout-aware polling and retry scenarios, including rate-limit and backoff edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/teng-lin/notebooklm-py/pull/1147?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/_artifact_polling.py [hunk 1]
```diff
@@ -295,28 +314,28 @@ async def _run_poll_loop(
...
                 logger.warning(
                     "wait_for_completion: transient %s on poll #%d, retrying in %.1fs",
                     e.__class__.__name__,
                     poll_retry_count,
                     backoff,
                 )
-                await asyncio.sleep(backoff)
+                await self._resolve_sleep()(backoff)
+                if deadline.expired():
+                    raise
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_artifact_polling.py]: Addressed in 55df6cf689966836f8e168eef004554f0bef58c5: transient polling retries now raise the artifact timeout diagnostic error when the remaining deadline is exhausted, chaining the transient exception as the cause. Also restored the prior poll-before-timeout diagnostics behavior and kept zero-interval immediate repolls working.

**Comment by teng-lin** [comment]
@claude review
